containers-tests/tests/map-strictness.hs

```diff
@@ -26,14 +42,887 @@ instance (Arbitrary k, Arbitrary v, Ord k) =>
          Arbitrary (Map k v) where
     arbitrary = M.fromList `fmap` arbitrary
 
+instance (Arbitrary a, Ord a) => Arbitrary (Set a) where
+  arbitrary = Set.fromList <$> arbitrary
+
```

Comment:
This is no good. For our purposes here, we want to build all `Set` shapes, as we do in `Set` testing, so we can make sure the implemented balance invariants actually match properly. Best to factor out a module for the existing instance, or perhaps to write a more general, less efficient version of it.

Reply:
I agree on this and I tried to move the existing instances, but they have a weird `IsInt` constraint which means they cannot generate `OrdA`.

https://github.com/haskell/containers/blob/fbade40798546a7fb7cbabed1922ef7b0010ab1f/containers-tests/tests/set-properties.hs#L219

https://github.com/haskell/containers/blob/fbade40798546a7fb7cbabed1922ef7b0010ab1f/containers-tests/tests/map-properties.hs#L294

Reply:
We can port the shape generation without the performance-only type restriction. Generate a shape and a list to match.

Reply:
I'm nervous about putting in tests that don't test everything they seem to.

Reply:
Alright, I'll take a look. I haven't attempted to understand the tree generation and the constraint.

By the way, why is it not simply `Arbitrary a => Arbitrary (Set a)`?

Reply:
> We can port the shape generation without the performance-only type restriction. Generate a shape and a list to match.

Done.

> Okay, so ... it totally _could_ be `(Arbitrary a, Ord a) => Arbitrary (Set a)`. The only reason it isn't is that it's faster to generate `Int`s like that than to generate an ordered list of unique arbitrary items. It might be worth seeing how much slower test performance is with a more general instance. Test performance does matter, because the faster the individual tests run the more we can use without driving contributors crazy with CI times.

While I agree that we should care about test performance, this might be worrying out of proportion. If you look at any CI run, say [this one](https://github.com/haskell/containers/actions/runs/10474461204/job/29008781350), you will see that running tests takes only 17s out of the total 9m 50s.